addons/account/models/account_payment.py

```diff
@@ -332,18 +337,22 @@ def _compute_show_require_partner_bank(self):
             payment.require_partner_bank_account = payment.state == 'draft' and payment.payment_method_code in self._get_method_codes_needing_bank_account()
 
     @api.depends('partner_id', 'company_id', 'payment_type')
-    def _compute_partner_bank_id(self):
-        ''' The default partner_bank_id will be the first available on the partner. '''
+    def _compute_available_partner_bank_ids(self):
         for pay in self:
             if pay.payment_type == 'inbound':
                 bank_partner = pay.company_id.partner_id
             else:
                 bank_partner = pay.partner_id
+            pay.available_partner_bank_ids = bank_partner.bank_ids\
+                    .filtered(lambda x: x.company_id.id in (False, pay.company_id.id))._origin
```

Comment:
Bug:

- Create 2 bank accounts A and B for your company, linked to journals J1 and J2
- open the form to create a new account.payment
- make partner_bank_id field visible in the form view (just to see the problem)
- click payment type = 'receive money'
- change journal (from J1 to J2)
- recipient bank account is not recomputed. It's still A, should be B.

(and if you fix it manually and change the partner, B is replaced by A again)

---------------

=> This field should only accept the bank set on the bank journal when receiving money (so, the journal should also be part of the dependencies)